### ISSUE TITLE
mongodb: Incrementally drop temp collections.

### ISSUE THREAD

**Issue by sellout** [quasar-analytics/quasar#540]
Temp collections can get large, and sometimes we make a few of them. And if there's an error along the way, we never get around to cleaning them up at the end.

It would nice to delete each temp collection once it's no longer needed, rather than waiting until the end of the request processing.

**Comment by mossprescott** [comment]
Yes, this is clearly needed, especially when working on the planner.

It gets complicated because temp collections are created by one step and become garbage when another step is completed. Effectively, they're in scope for some portion of the plan. As plans get more complex (i.e. joins), these scopes become nested. Eventually, we might want to execute steps in parallel. And so on.

On the other hand, when we generate JS plans for Mongo, we need to boil the plan down into a sequence of commands. To cleanup in the face of failure, we'll need to generate JS `try/finally` blocks, assuming that works in the mongo shell (and other tools).

So I think what we need here is an ADT for plans that captures the scopes and associated temp collections. We would interpret it against the MongoDB API, and compile it to JS with try/finally.

**Comment by jdegoes** [comment]
If we add a "side-effecting" delete collection to the workflow task (something like `DeleteCollection(c: Collection, source: WorkflowTask)`), then the evaluator could remain totally dumb, just blindly following instructions created by a much smarter planner. The challenge would be in constructing a sane way of building self-cleaning up workflows, which could probably done with combinators. 

Will think on it more...
